Changes to make pycbc_make_coinc_search_workflow flexible to reading in a template bank in both hdf and xml format

soumide1102 opened the pull request:
This PR is about making `pycbc_make_coinc_search_workflow` flexible to be able to read in the template bank in both hdf/h5 and xml format. If the bank read in is in hdf format, it is used as it is in the workflow. If not in hdf/h5 format (like xml), it is converted to hdf format and then used in the full data and injections jobs.

spxiwh commented on bin/workflows/pycbc_make_coinc_search_workflow:
```diff
@@ -155,7 +155,23 @@ precalc_psd_files = wf.setup_psd_workflow(workflow, analyzable_segs,
+    hdfbank = bank_files
+else :
+    hdfbank = wf.convert_bank_to_hdf(workflow, bank_files, "bank")
+logging.info("Input bank is a %s file", ext)
```
Do you think it's possible to move all this into setup_tmpltbank_workflow? Maybe add a return_format key-word argument. If not given the bank just comes back in whatever format it is supplied in, if hdf, it does what you've added here, if XML it either returns an XML file (if we have one) or fails if we have something else. If value is given and is *not* XML or HDF then fail. This then allows this to be easily used in other applications.

ahnitz replied:
If you decide to add a return format keyword, the workflow should choose 'hdf' internally. Let's drop xml for anything but the very first input stage.

soumide1102 replied:
@spxiwh If I understand it correctly, do you want it to do the part of checking what the extension of the bank is in `setup_tmpltbank_workflow` , and then if the bank is in hdf format, return as it is, if it is in xml call `convert_bank_to_hdf ` to convert it to hdf in `setup_tmpltbank_workflow` , and if any other format, just return an error.. ie. do all of the changes in `setup_tmpltbank_workflow` ?

spxiwh replied:
@soumide1102 Yeah, let me give the motivation here, but mostly what you say is true:

I would like to have all the logic (that you describe) within the `setup_tmpltbank_workflow` function. Other workflows also use that function so having this code in one place avoids potential duplication of code.

*However* PyGRB workflow also uses this function, and that is still using XML template banks, so I would like to have the ability to choose whether the bank is returned in XML or HDF format (or potentially other formats in the future). (If the input bank is in HDF format and the user asks for XML, or if a third format is specified anywhere, just raise a NotImplementedError ... If someone wants to use that code path, they can implement it!)

ahnitz commented on bin/workflows/pycbc_make_coinc_search_workflow:
```diff
@@ -155,7 +155,23 @@ precalc_psd_files = wf.setup_psd_workflow(workflow, analyzable_segs,
+
+# Determine the extension of the bank file taken as input
+cp = workflow.cp
+tmpltbankMethod = cp.get_opt_tags("workflow-tmpltbank", "tmpltbank-method", '')
```
Seems a bit overkill to look at the configparser here. An easier idea is to simply check the output of the bank_files to see what the file extension is. You can get it from the logical file name. You should be able to get this from something like bank_files[0].name.

soumide1102 replied:
@ahnitz Thanks. I'll change that.

soumide1102 commented:
@spxiwh I've moved the format conversion of the input bank to `setup_tmpltbank_workflow` , and added a keyword argument to the function to specify the format of the bank to be returned, as you wanted. It should now be able to return the bank in a format as per the user's choice. Could you please take a look? 

Landscape is however complaining about `setup_tmpltbank_workflow` having too many local variables and its complexity, and I wonder how to deal with that issue if fixing it is necessary.

spxiwh commented:
Hi @soumide1102. Great, thanks for taking care of my concerns. Unless I hear from Alex in the next 20 minutes I'll merge this.

... Please note that I added `.hdf5` as a valid extension for HDF files, it matches what sbank is doing.

spxiwh commented:
I also agree to ignore the "complexity" warnings. These are telling us that we are doing too many operations in one function and should use more sub-functions/classes/methods. Some restructuring is needed to deal with this, or we just raise the threshold for these warnings.

soumide1102 commented:
Thanks @spxiwh !
